joomla/joomla-cms
# Don't append but set the "Content-Encoding" header value in htaccess.txt

Pull Request for Issue #39176 .

### Summary of Changes

The `Header append Content-Encoding gzip` is changed to `Header set Content-Encoding gzip` in the section for the gzipped resource files of the `htaccess.txt` file.

This makes sure not to get an invalid value "gzip gzip" value but a correct "gzip" for the "Content-Encoding" header of the gzipped resource files when following conditions are met:
- Apache web server with mod_rewrite in use
- mod_gzip or mod_deflate are not enabled (otherwise you get double compression anyway).
- Joomla is installed in a subdirectory of your domain or your local webserver's document root.
- An `.htaccess` file is used in that subdirectory which is based on the Joomla 4 `htaccess.txt` file and so has the last section for the gzipped resource files in use.
- The gzipped resource files are present, i.e. you are not on a Git clone but on a real Joomla installation made with an installation package (on a Git clone you can run `npm run gzip` to create the gzipped resources for testing).
- In the parent directory, i.e. the domain directory or your local webserver's document root is also an `.htaccess` file like that in use. That can be the case e.g. if you have the above mentioned Joomla installation in a subdirectory of a Joomla installation.

See also the Drupal issue https://www.drupal.org/project/drupal/issues/1116416 linked in #39176 for a detailed description and discussion of the issue.

As concluded there in the discussion, the initial approach to do the same change also for the later `Header append Vary Accept-Encoding` line was abandoned because that header value can be a list of multiple http headers, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary . Therefore I also don't change that in our `htaccess.txt`.

In addition, a postinstall message is added, but only for updated sites, not for new installations, as the latter come already with the correct htaccess.txt file.

### Testing Instructions

#### 1. Testing the .htaccess change

1. Have an Apache webserver with mod_rewrite enabled and mod_gzip and mod_deflate disabled.

2. Have an `.htaccess` file in your webserver's document root which contains at least the last section of our Joomla 4 `htaccess.txt` file:
```
<IfModule mod_headers.c>
	# Serve gzip compressed CSS files if they exist
	# and the client accepts gzip.
	RewriteCond "%{HTTP:Accept-encoding}" "gzip"
	RewriteCond "%{REQUEST_FILENAME}\.gz" -s
	RewriteRule "^(.*)\.css" "$1\.css\.gz" [QSA]

	# Serve gzip compressed JS files if they exist
	# and the client accepts gzip.
	RewriteCond "%{HTTP:Accept-encoding}" "gzip"
	RewriteCond "%{REQUEST_FILENAME}\.gz" -s
	RewriteRule "^(.*)\.js" "$1\.js\.gz" [QSA]

	# Serve correct content types, and prevent mod_deflate double gzip.
	RewriteRule "\.css\.gz$" "-" [T=text/css,E=no-gzip:1]
	RewriteRule "\.js\.gz$" "-" [T=text/javascript,E=no-gzip:1]

	<FilesMatch "(\.js\.gz|\.css\.gz)$">
		# Serve correct encoding type.
		Header append Content-Encoding gzip

		# Force proxies to cache gzipped &
		# non-gzipped css/js files separately.
		Header append Vary Accept-Encoding
	</FilesMatch>
</IfModule>
```
This can be because you already have a Joomla 4 installation there and you use URL rewriting.

3. Create a new subdirectory below that and install a Joomla 4.2.5 or latest 4.2.6-dev nightly build into that directory.

4. Enable URL rewriting for that installation and copy `htaccess.txt` to `.htaccess`.

5. Request some page of that installation with a forced reload and watch the network tab of your browser's developer tools.

6. Check the "Content-Encoding" in the http response of one of the gzipped css or js resources.
Result: Value "gzip gzip", see section "Actual result BEFORE applying this Pull Request" below.

7. Change the `.htaccess` file of the installation in the subdirectory in the same way as this PR changes the `htaccess.txt` file.

8. Repeat steps 4 and 5.
Result: Value "gzip", see section "Expected result AFTER applying this Pull Request" below.

#### 2. Testing the postinstall message

1. Make a new installation with the installation package created by drone for this branch and verify that the new postinstall message is not shown.

2. Update a 4.2.8 or older or a current 4.2-dev branch to the update package created by drone for this PR and verify that after the update the new postinstall message is shown.

### Actual result BEFORE applying this Pull Request

Header "Content-Encoding" has value "gzip gzip" in the http response for a gzipped resource file.

### Expected result AFTER applying this Pull Request

Header "Content-Encoding" has value "gzip" in the http response for a gzipped resource file.

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Preconfigured_htaccess


When the PR gets merged, I will care for the update of the documentation mentioned above.
